dev/src/bundle.ts

```diff
@@ -0,0 +1,225 @@
+  ) {
+    if (docOrName instanceof DocumentReference) {
+      this.docRefs.push(docOrName);
+    } else if (docOrName instanceof DocumentSnapshot) {
+      this.docSnaps.push(docOrName);
+    } else if (query instanceof Query) {
+      this.queries.set(docOrName, query);
+    } else if (query instanceof QuerySnapshot) {
+      this.querySnaps.set(docOrName, query);
+    }
+
+    return this;
+  }
+
+  private addBundledDocument(snap: DocumentSnapshot) {
+    if (snap.exists) {
+      const docProto = snap.toDocumentProto();
+      if (
+        !this.documents.has(snap.id) ||
+        Timestamp.fromProto(this.documents.get(snap.id)!.metadata.readTime!) <
+          snap.readTime
+      ) {
+        this.documents.set(snap.id, {
+          document: docProto,
+          metadata: {
+            documentKey: docProto.name,
+            readTime: snap.readTime.toProto().timestampValue,
+          },
+        });
+      }
+      if (snap.readTime > this.latestReadTime) {
+        this.latestReadTime = snap.readTime;
+      }
+    }
+  }
+
+  private addNamedQuery(name: string, querySnap: QuerySnapshot) {
+    this.namedQueries.set(name, {
+      name,
+      bundledQuery: querySnap.query.toBundledQuery(),
+      readTime: querySnap.readTime.toProto().timestampValue,
+    });
+
+    for (const snap of querySnap.docs) {
+      this.addBundledDocument(snap);
+    }
+
+    if (querySnap.readTime > this.latestReadTime) {
+      this.latestReadTime = querySnap.readTime;
+    }
+  }
+
+  private pushLengthPrefixedString(readable: Readable, payload: string) {
+    const buffer = Buffer.from(payload, 'utf-8');
+    const lengthBuffer = Buffer.from(buffer.length.toString());
+
+    readable.push(Buffer.concat([lengthBuffer, buffer]));
+  }
+
+  stream(): NodeJS.ReadableStream {
+    assert(
+      this.docRefs.length +
+        this.docSnaps.length +
+        this.queries.size +
+        this.querySnaps.size >
+        0,
+      'Nothing is added to the bundle.'
+    );
+    for (const snap of this.docSnaps) {
+      this.addBundledDocument(snap);
+    }
+
+    for (const [name, snap] of Array.from(this.querySnaps)) {
+      if (this.namedQueries.has(name)) {
+        throw new Error(`Query name conflict: ${name}`);
+      }
+
+      this.addNamedQuery(name, snap);
+    }
+
+    const promises: Array<Promise<void>> = [];
+    for (const doc of this.docRefs) {
+      promises.push(
+        doc.get().then(snap => {
+          this.addBundledDocument(snap);
+        })
+      );
+    }
+
+    for (const [name, query] of Array.from(this.queries)) {
+      if (this.namedQueries.has(name)) {
+        throw new Error(`Query name conflict: ${name}`);
+      }
+
+      promises.push(
+        query.get().then(snap => {
+          const querySnap = snap as QuerySnapshot;
+          this.addNamedQuery(name, querySnap);
+        })
+      );
+    }
+
+    const readable = new Readable({
+      objectMode: false,
+      read: async size => {
```

Comment:
Read cannot be `async` as far as I can tell. It needs to return the data that is being read.

Reply:
Note this is the `read` from `ReadableOptions`, and its return type is a `void`.

However I will switch to asyncIterator anyways, so it does not matter any more.